### PR TITLE
Outside passer/cube chart 3

### PR DESCRIPTION
**Pull request by outsidePasser**
Fixes #[PUT_ISSUE_NUMBER_HERE]

> Please minimize the amount of changes to shared `lib/engine` code, if possible

> If you are implementing a new game, please break up the changes into multiple PRs for ease of review.

### Before clicking "Create"

- [ ] Branch is derived from the latest `master`
- [ ] Add the `pins` label if this change will break existing games
- [ ] Code passes linter with `docker compose exec rack rubocop -a`
- [ ] Tests pass cleanly with `docker compose exec rack rake`

### Implementation Notes

* **Explanation of Change**

* **Screenshots**

* **Any Assumptions / Hacks**

### CODE REVIEW

File: assets/app/view/game/loan_chart.rb [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        contents = []
+        contents << h(:div, { style: name_style }, @game.loan_entity_name)
+        contents << h(:div, { style: grid_style }, chart)
+        h(:div, contents)
+      end
+
+      def chart
+        contents = []
+        @loan_chart.each do |row|
+          contents << render_first_cell(row[0])
```
Comment (crericha) [inline assets/app/view/game/loan_chart.rb]: Is there a better way we can denote the row header? It's not intuitive.

File: assets/app/view/game/loan_chart.rb [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+          row.drop(1).each do |cell|
+            contents << render_cube_cell(cell, @empty_cube_spaces.zero?)
```
Comment (crericha) [inline assets/app/view/game/loan_chart.rb]: Suggest making the names less about the CSS structure and more about what they represent. Here it would be render_loan.

File: assets/app/view/game/loan_chart.rb [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+            @empty_cube_spaces -= 1 if @empty_cube_spaces.positive?
```
Comment (crericha) [inline assets/app/view/game/loan_chart.rb]: I would just have each loan be nil or a loan object, so you don't have to count. Nor do the loans have to be in order (if for some reason that happens in a future game).

File: assets/app/view/game/loan_chart.rb [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+          end
+          (@num_columns - row.size).times { contents << render_empty_cell }
```
Comment (crericha) [inline assets/app/view/game/loan_chart.rb]: Can this be in render_loan? If note there, its empty. If there it has something.

File: assets/app/view/game/loan_chart.rb [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+      def render_first_cell(label)
+        @stock_movement_style = {
+          display: 'flex',
+          justifyContent: 'center',
+          alignItems: 'center',
+          fontSize: 'xx-large',
+          backgroundColor: 'gray',
+        }
+
+        h(:div, { style: @stock_movement_style }, label)
```
Comment (crericha) [inline assets/app/view/game/loan_chart.rb]: Only use @ if you need the variable outside of the method. Otherwise just use a local variable.

File: assets/app/view/game/loan_chart.rb [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+          paddingBottom: '5px',
+        }
+
+        cube_props = {
+          attrs: {
+            src: '../icons/red_cube.svg',
+            title: 'cube',
+          },
+          style: {
+            objectFit: 'scale-down',
```
Comment (crericha) [inline assets/app/view/game/loan_chart.rb:90]: Interesting. I'll have to look objectFit up. Never heard of it.

File: assets/app/view/game/player.rb [hunk 7]
```diff
@@ -142,6 +142,13 @@ def render_info
           ])
         end
 
+        if @game.respond_to?(:player_loans) && @game.player_loans(@player).positive?
```
Comment (crericha) [inline assets/app/view/game/player.rb]: We don't want to show 0 / X loans for players that haven't taken a loan?

File: assets/app/view/game/round/stock.rb [hunk 8]
```diff
@@ -93,7 +93,8 @@ def render_buttons
           buttons = []
           buttons.concat(render_merge_button) if @current_actions.include?('merge')
           buttons.concat(render_payoff_player_debt_button) if @current_actions.include?('payoff_player_debt')
-
+          buttons.concat(render_take_loan) if @current_actions.include?('take_loan')
```
Comment (crericha) [inline assets/app/view/game/round/stock.rb:96]: How does this interact with 1817 which has a take loan action in the stock round also.

File: assets/app/view/game/round/stock.rb [hunk 9]
```diff
@@ -404,6 +405,20 @@ def render_select_par_slot
 
           h(:div, children)
         end
+
+        def render_take_loan
+          take_loan = lambda do
+            process_action(Engine::Action::TakeLoan.new(@current_entity, loan: nil))
+          end
+          [h(:button, { on: { click: take_loan } }, 'Take Loan')]
```
Comment (crericha) [inline assets/app/view/game/round/stock.rb]: Can we add the loan amount?

File: assets/app/view/game/round/stock.rb [hunk 10]
```diff
@@ -404,6 +405,20 @@ def render_select_par_slot
...
+        end
+
+        def render_payoff_loan
+          payoff_loan = lambda do
+            process_action(Engine::Action::PayoffLoan.new(@current_entity, loan: nil))
+          end
+          [h(:button, { on: { click: payoff_loan } }, 'Payoff Loan')]
```
Comment (crericha) [inline assets/app/view/game/round/stock.rb]: Can we add the payoff amount?

File: assets/app/view/game/stock_market.rb [hunk 11]
```diff
@@ -225,10 +234,11 @@ def grid_zigzag(zigzag)
...
 
-          element = h(:div, { style: cell_style(box_style, price.types) }, [
-                      h(:div, { style: PRICE_STYLE_1D }, price.price),
-                      h(:div, { style: TOKEN_STYLE_1D }, tokens),
-                    ])
+          cell_elements = [h(:div, { style: PRICE_STYLE_1D }, price.price),
+                           h(:div, { style: TOKEN_STYLE_1D }, tokens)]
+          cell_elements << h(:div, { style: PRICE_STYLE_INFO }, price.info) if price.info
+
+          element = h(:div, { style: cell_style(box_style, price.types) }, cell_elements)
```
Comment (crericha) [inline assets/app/view/game/stock_market.rb:241]: Why was this changed?

File: lib/engine/game/base.rb [hunk 12]
```diff
@@ -212,6 +212,7 @@ class Base
 
       MINORS = [].freeze
 
+      PLAYER_CLASS = Player
```
Comment (crericha) [inline lib/engine/game/base.rb]: Is there some place better to group this?

File: lib/engine/game/g_1877_stockholm_tramways/step/acquire.rb [hunk 13]
```diff
@@ -115,7 +115,7 @@ def process_buy_shares(action)
 
             # buy share for half its price
             share_price = corporation.share_price
-            corporation.share_price = SharePrice.new(share_price.coordinates, price: share_price.price / 2)
+            corporation.share_price = SharePrice.new(share_price.coordinates, price: share_price.price / 2, info: nil)
```
Comment (crericha) [inline lib/engine/game/g_1877_stockholm_tramways/step/acquire.rb]: You don't need to pass in unused kwargs. They are optional.

File: lib/engine/game/g_18_eus/game.rb [hunk 14]
```diff
@@ -24,12 +29,19 @@ class Game < Game::Base
...
-        ].freeze
+        # MARKET = [
+        #  ["40", "44", "47", "50p", "53p", "57p", "61p", "65p", "70p", "75p", ["80p", "12"], ["86p", "12"],
+        #   ["92p", "14"], ["98p", "14"], ["105x", "16"], ["112x", "16"], ["120x", "18"], ["128x", "18"],
+        #   ["137x", "20"], ["147x", "20"], ["157x", "22"], ["168z", "22"], ["180z", "24"], ["193z", "26"],
+        #   ["206z", "28"], ["221", "30"], ["236", "32"], ["253", "34"], ["270", "38"], ["289", "40"],
+        #   ["310", "40"], ["331", "40"], ["354", "40"], ["379", "50"], ["406k", "50"], ["434k", "50"],
+        #   ["465k", "50"], ["497k", "60"], ["532k", "60"], ["569k", "60"], ["609k", "70"], ["652k", "70"],
+        #   ["700k", "80"], "750e", "800e"]
+        # ].freeze
```
Comment (crericha) [inline lib/engine/game/g_18_eus/game.rb]: Remove commented out code

File: lib/engine/game/g_18_eus/game.rb [hunk 15]
```diff
@@ -472,6 +485,132 @@ def company_status_str(company)
         def bidding_token_per_player
           self.class::BIDDING_BOX_PRIVATE_COUNT
         end
+
+        def operating_order
+          super.partition { |corp| corp != bny }.flatten
```
Comment (crericha) [inline lib/engine/game/g_18_eus/game.rb:476]: Is partition guaranteed to return the non-bank partition first? I worry this is working because we are lucky.

File: lib/engine/game/g_18_eus/game.rb [hunk 16]
```diff
@@ -472,6 +485,132 @@ def company_status_str(company)
...
+                { 'stock_movement' => :diagonal_and_straight, 'multipliers' => [5, 5, 5, 5, 5, 5, 5, 5, 5, 5] },
+              ]
+            end
+          @loans_taken = 0
+        end
+
+        def loan_chart
+          loan_chart = []
+          @loans.each do |row|
+            loan_chart << ([translate_loan_movement(row['stock_movement'])] + row['multipliers']).flatten
```
Comment (crericha) [inline lib/engine/game/g_18_eus/game.rb]: Maybe [{header: translate_loan_movement..., loans: }]. Just an idea.

File: lib/engine/game/g_18_eus/game.rb [hunk 17]
```diff
@@ -472,6 +485,132 @@ def company_status_str(company)
...
+          end
+          loan_chart
+        end
+
+        def translate_loan_movement(movement)
```
Comment (crericha) [inline lib/engine/game/g_18_eus/game.rb]: loan_movement_to_arrows?

File: lib/engine/game/g_18_eus/game.rb [hunk 18]
```diff
@@ -472,6 +485,132 @@ def company_status_str(company)
...
+
+        def current_loan_values
+          return [:none, 0] if @loans_taken.zero?
+
+          taken = @loans_taken
+          @loans.each do |row|
+            return [row['stock_movement'], row['multipliers'][taken - 1]] if taken <= row['multipliers'].size
+
+            taken -= row['multipliers'].size
+          end
```
Comment (crericha) [inline lib/engine/game/g_18_eus/game.rb]: This feels pretty intense to figure out the current loan value. Can we calculate on take loan and payoff loan? Also, shouldn't it be multiplier, not value?

File: lib/engine/game/g_18_eus/game.rb [hunk 19]
```diff
@@ -472,6 +485,132 @@ def company_status_str(company)
...
+          else 10
+          end
+        end
+
+        def player_loans(player)
+          player.loans
+        end
+
+        def can_take_loan?(player)
+          player.loans < max_player_loans && !bny.player_share_holders.include?(player)
```
Comment (crericha) [inline lib/engine/game/g_18_eus/game.rb]: I think player_share_holders includes people who sold their share. Make sure this works where you can take a loan in one SR and sell in another and take a loan.

File: lib/engine/game/g_18_eus/market.rb [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+
+require_relative 'map'
+
+module Engine
+  module Game
+    module G18EUS
+      module Market
+        MARKET = [
+          [
+            { price: 40, types: [] },
```
Comment (crericha) [inline lib/engine/game/g_18_eus/market.rb]: Don't include types if it is empty. We should support not having types when the share box doesn't need it.

File: lib/engine/game/g_18_eus/round/stock.rb [hunk 21]
```diff
@@ -7,11 +7,13 @@ module Game
     module G18EUS
       module Round
         class Stock < Engine::Round::Stock
-          attr_accessor :bids
+          attr_accessor :bids, :take_loan_list, :payoff_loan_list
```
Comment (crericha) [inline lib/engine/game/g_18_eus/round/stock.rb]: I'm personally not a fan of adding the container style to the name. I prefer something like taken_loans, paid_loans.

File: lib/engine/game/g_18_eus/step/buy_sell_par_shares.rb [hunk 22]
```diff
@@ -76,6 +78,20 @@ def modify_purchase_price(bundle)
...
             super
           end
+
+          def process_buy_shares(action)
+            return super unless action.bundle.corporation == @game.bny
+
+            @round.bought_from_ipo = true if action.bundle.owner.corporation?
+            buy_shares(action.entity, action.bundle, swap: action.swap, allow_president_change: false)
+            track_action(action, action.bundle.corporation)
+            log_pass(action.entity)
```
Comment (crericha) [inline lib/engine/game/g_18_eus/step/buy_sell_par_shares.rb]: I don't understand the need for all of this. Can you please explain?

File: lib/engine/game/g_18_eus/step/dividend.rb [hunk 23]
```diff
@@ -24,6 +37,50 @@ def share_price_change(entity, revenue = 0)
...
+
+            @game.players.each do |player|
+              next unless player.loans.positive?
+
+              player.spend(interest * player.loans, @game.bank)
+              @log << "#{player.name} pays #{@game.format_currency(interest * player.loans)} in interest " \
+                      "(#{@game.format_currency(interest)} per share)"
+            end
+
+            payout_shares(action.entity, interest * multiplier * 10)
```
Comment (crericha) [inline lib/engine/game/g_18_eus/step/dividend.rb]: If a player has loans and the interest payment private, we have to handle the situation where their cash goes negative before the privates pays them and makes them positive again.

File: lib/engine/game/g_18_eus/step/dividend.rb [hunk 24]
```diff
@@ -24,6 +37,50 @@ def share_price_change(entity, revenue = 0)
           def movement_str(times, dir)
             "#{times / 2} #{dir}"
           end
+
+          def corporation_dividends(entity, per_share)
+            entity == @game.bny ? 0 : super
+          end
+
+          def process_dividend(action)
+            return super unless action.entity == @game.bny
```
Comment (crericha) [inline lib/engine/game/g_18_eus/step/dividend.rb]: I'd like to see this method use super all the time for the payout part by making @game.routes_revenue(routes) return the payout amount. If this works, less duplicate code that can grow stale.

We'd do loan payments first and then super.

File: lib/engine/game/g_18_eus/step/dividend.rb [hunk 25]
```diff
@@ -24,6 +37,50 @@ def share_price_change(entity, revenue = 0)
...
+              player.spend(interest * player.loans, @game.bank)
+              @log << "#{player.name} pays #{@game.format_currency(interest * player.loans)} in interest " \
+                      "(#{@game.format_currency(interest)} per share)"
+            end
+
+            payout_shares(action.entity, interest * multiplier * 10)
+
+            spaces = stock_movement_to_spaces(movement)
+            if spaces.positive?
+              spaces.times { @game.stock_market.move_right(@game.bny) }
```
Comment (crericha) [inline lib/engine/game/g_18_eus/step/dividend.rb]: Move "right" is incorrect. "up" is diagonal.

See if you can use change_share_price() here as part of integrating with the common code.

File: lib/engine/game/g_18_eus/step/dividend.rb [hunk 26]
```diff
@@ -24,6 +37,50 @@ def share_price_change(entity, revenue = 0)
...
+              @log << "BNY moves #{spaces} diagonal space(s) to the right to " \
+                      "#{@game.format_currency(@game.bny.share_price.price)}"
+            end
+
+            if @game.bny.num_treasury_shares < 10
+              @game.stock_market.move_right(@game.bny)
```
Comment (crericha) [inline lib/engine/game/g_18_eus/step/dividend.rb]: move_up

File: lib/engine/game/g_18_eus/step/loan_taker.rb [hunk 27]
```diff
@@ -0,0 +1,39 @@
+# frozen_string_literal: true
+
+module Engine
+  module Game
+    module G18EUS
+      module Step
+        module LoanTaker
+          def actions(entity)
+            actions = []
```
Comment (crericha) [inline lib/engine/game/g_18_eus/step/loan_taker.rb]: I think you need to check for entity == current_entity

File: lib/engine/game/g_18_eus/step/loan_taker.rb [hunk 28]
```diff
@@ -0,0 +1,39 @@
...
+            actions << 'take_loan' if can_take_loan?(entity)
+            actions << 'payoff_loan' if can_payoff_loan?(entity)
+            actions.concat(super).uniq
+          end
+
+          def can_take_loan?(entity)
+            !bought? && entity.player? && !@round.payoff_loan_list.include?(entity) && @game.can_take_loan?(entity)
```
Comment (crericha) [inline lib/engine/game/g_18_eus/step/loan_taker.rb]: Also cannot take if the player owns a bank share. Is that checked in game?

File: lib/engine/game/g_18_eus/step/skip_bny.rb [hunk 29]
```diff
@@ -0,0 +1,17 @@
...
+
+module Engine
+  module Game
+    module G18EUS
+      module SkipBny
+        def actions(entity)
+          return [] if @game.bny == entity
+
+          super
+        end
```
Comment (crericha) [inline lib/engine/game/g_18_eus/step/skip_bny.rb:11]: Add log_skip here too and silence for BNY

File: lib/engine/stock_market.rb [hunk 30]
```diff
@@ -12,11 +12,18 @@ def initialize(market, unlimited_types, multiple_buy_types: [], zigzag: nil)
       @zigzag = zigzag
       @market = market.map.with_index do |row, r_index|
         row.map.with_index do |code, c_index|
-          price = SharePrice.from_code(code,
-                                       r_index,
-                                       c_index,
-                                       unlimited_types,
-                                       multiple_buy_types: multiple_buy_types)
+          price = if code.instance_of?(Hash)
+                    SharePrice.new([r_index, c_index],
```
Comment (crericha) [inline lib/engine/stock_market.rb]: SharePrice.new([r_index, c_index], **code)

Comment (crericha) [reply lib/engine/stock_market.rb]: I believe ** is the hash expansion. It is for python for sure and I think that's the universal syntax. If not, there *is* a syntax for it where you don't have to specify each by hand.

File: lib/engine/stock_market.rb [hunk 31]
```diff
@@ -12,11 +12,18 @@ def initialize(market, unlimited_types, multiple_buy_types: [], zigzag: nil)
...
+                                   price: code[:price],
+                                   types: code[:types],
+                                   info: code[:info])
```
Comment (crericha) [inline lib/engine/stock_market.rb]: Don't forget unlimited_types and multiple_buy_types:

File: lib/engine/game/g_18_eus/step/loan_taker.rb (hunk 28 above)
Comment (outsidePasser) [reply lib/engine/game/g_18_eus/step/loan_taker.rb]: That is checked in game:

        def can_take_loan?(player)
          player.loans < max_player_loans && !bny.player_share_holders.include?(player)
        end
        
 I checked those things that were not dependent on the state of the round in game.

File: assets/app/view/game/loan_chart.rb (hunk 5 above)
Comment (outsidePasser) [reply assets/app/view/game/loan_chart.rb]: Changed

File: assets/app/view/game/player.rb (hunk 7 above)
Comment (outsidePasser) [reply assets/app/view/game/player.rb]: I'm fine with doing that - most of the things in the player don't display unless they are applicable today.

File: assets/app/view/game/round/stock.rb (hunk 8 above)
Comment (outsidePasser) [reply assets/app/view/game/round/stock.rb:96]: In 1817 the "Take Loan" happens inside a corporation.  This is at the top.

File: assets/app/view/game/stock_market.rb (hunk 11 above)
Comment (outsidePasser) [reply assets/app/view/game/stock_market.rb:241]: This was changed to add in the price.info if it exists.

File: lib/engine/game/g_18_eus/game.rb (hunk 14 above)
Comment (outsidePasser) [reply lib/engine/game/g_18_eus/game.rb]: Removed

File: lib/engine/game/g_18_eus/game.rb (hunk 18 above)
Comment (outsidePasser) [reply lib/engine/game/g_18_eus/game.rb]: We can, but it's the same intensity at that point.  

This returns both the multiplier and the stock movement - so it's maybe not "values".  I could probably come up wither a better name.

File: lib/engine/game/g_18_eus/game.rb (hunk 19 above)
Comment (outsidePasser) [reply lib/engine/game/g_18_eus/game.rb]: I think you may be right - I think I complained about this in a previous game implementation, and was told that it was "working as intended", even though it isn't very evident.  I'll look into it.

File: lib/engine/game/g_18_eus/step/skip_bny.rb [hunk 32]
```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+
+require_relative '../../../step/buy_company'
```
Comment (crericha) [inline lib/engine/game/g_18_eus/step/skip_bny.rb]: I don't think the require is needed.

File: lib/engine/stock_market.rb (hunk 30 above)
Comment (outsidePasser) [reply lib/engine/stock_market.rb]: Changed

File: lib/engine/stock_market.rb (hunk 31 above)
Comment (outsidePasser) [reply lib/engine/stock_market.rb]: Fixed

File: lib/engine/game/g_18_eus/game.rb (hunk 15 above)
Comment (outsidePasser) [reply lib/engine/game/g_18_eus/game.rb:476]: It is guaranteed - it returns those for which the block assess to "true" first.

File: assets/app/view/game/loan_chart.rb (hunk 1 above)
Comment (outsidePasser) [reply assets/app/view/game/loan_chart.rb]: Changed to make it clearer.  New row structure is: r = {"header"=>"↗", "loans"=>[[{"value"=>0.5, "loan_taken"=>true}], ...

File: assets/app/view/game/loan_chart.rb (hunk 2 above)
Comment (outsidePasser) [reply assets/app/view/game/loan_chart.rb]: Changed to render_row_header and render_loan

File: assets/app/view/game/loan_chart.rb (hunk 3 above)
Comment (outsidePasser) [reply assets/app/view/game/loan_chart.rb]: Changed to have game put in a "nil" for an empty loan.

File: assets/app/view/game/loan_chart.rb (hunk 4 above)
Comment (outsidePasser) [reply assets/app/view/game/loan_chart.rb]: Moved to render_loan

File: assets/app/view/game/player.rb (hunk 7 above)
Comment (outsidePasser) [reply assets/app/view/game/player.rb]: Added

File: assets/app/view/game/round/stock.rb (hunk 9 above)
Comment (outsidePasser) [reply assets/app/view/game/round/stock.rb]: Added

File: assets/app/view/game/round/stock.rb (hunk 10 above)
Comment (outsidePasser) [reply assets/app/view/game/round/stock.rb]: Added

File: lib/engine/game/base.rb (hunk 12 above)
Comment (outsidePasser) [reply lib/engine/game/base.rb]: I didn't really see someplace that was an automatic "yes, it goes here".  I moved it to near TRAIN_CLASS and DEPOT_CLASS for now.

File: lib/engine/game/g_1877_stockholm_tramways/step/acquire.rb (hunk 13 above)
Comment (outsidePasser) [reply lib/engine/game/g_1877_stockholm_tramways/step/acquire.rb]: It was needed the way I did kwargs wrong before - no longer needed, and removed.

File: lib/engine/game/g_18_eus/game.rb (hunk 16 above)
Comment (outsidePasser) [reply lib/engine/game/g_18_eus/game.rb]: Changed to do it that way.

File: lib/engine/game/g_18_eus/game.rb (hunk 17 above)
Comment (outsidePasser) [reply lib/engine/game/g_18_eus/game.rb]: Changed

File: lib/engine/game/g_18_eus/game.rb (hunk 18 above)
Comment (outsidePasser) [reply lib/engine/game/g_18_eus/game.rb]: Came up with another way - instead of recalculating "row" and "col" all the time, built a map during setup so that I can just get the row and col from the current number of taken loans.

File: lib/engine/game/g_18_eus/step/dividend.rb (hunk 24 above)
Comment (outsidePasser) [reply lib/engine/game/g_18_eus/step/dividend.rb]: The main reason I broke this out is that the stock movement is different than what is done in process_dividends.  I'll take a look and see if I can do something about that.

File: lib/engine/game/g_18_eus/market.rb (hunk 20 above)
Comment (outsidePasser) [reply lib/engine/game/g_18_eus/market.rb]: Removed

File: lib/engine/game/g_18_eus/game.rb (hunk 19 above)
Comment (outsidePasser) [reply lib/engine/game/g_18_eus/game.rb]: Confirmed you are correct, changed to !(bny.player_share_holders[player]&.positive?)

File: lib/engine/game/g_18_eus/round/stock.rb (hunk 21 above)
Comment (outsidePasser) [reply lib/engine/game/g_18_eus/round/stock.rb]: Changed

File: lib/engine/game/g_18_eus/step/buy_sell_par_shares.rb (hunk 22 above)
Comment (outsidePasser) [reply lib/engine/game/g_18_eus/step/buy_sell_par_shares.rb]: This is there because in the base process_buy_shares, when buy_shares is called, it doesn't pass in allow_president_change which defaults to "true".  We want to that to be "false" for the bny.  

I suppose I could add allow_president_change to the base buy_shares action - didn't know if that was right approach, either.

Comment (outsidePasser) [reply lib/engine/game/g_18_eus/step/buy_sell_par_shares.rb]: Could also add a call out from process_buy_shares to @game.allow_president_change?(corporation) ?

File: lib/engine/game/g_18_eus/step/dividend.rb (hunk 23 above)
Comment (outsidePasser) [reply lib/engine/game/g_18_eus/step/dividend.rb]: Yes, we do, and I'm not sure how to do that.  Is this enough of a tails case that we don't need it for initial launch?

File: lib/engine/game/g_18_eus/step/dividend.rb (hunk 24 above)
Comment (outsidePasser) [reply lib/engine/game/g_18_eus/step/dividend.rb]: routes_revenue would work great if it passed in the running corporation so that I could check if it's the bny.  Is it worth adding that to call?

File: lib/engine/game/g_18_eus/step/buy_sell_par_shares.rb (hunk 22 above)
Comment (outsidePasser) [reply lib/engine/game/g_18_eus/step/buy_sell_par_shares.rb]: Added a call allow_president_change? to the step (instead of to @game).

File: lib/engine/game/g_18_eus/step/loan_taker.rb (hunk 27 above)
Comment (outsidePasser) [reply lib/engine/game/g_18_eus/step/loan_taker.rb]: Added

File: lib/engine/game/g_18_eus/step/skip_bny.rb (hunk 29 above)
Comment (outsidePasser) [reply lib/engine/game/g_18_eus/step/skip_bny.rb:11]: Added, and added steps that didn't have skip_bny yet

File: lib/engine/game/g_18_eus/step/skip_bny.rb (hunk 32 above)
Comment (outsidePasser) [reply lib/engine/game/g_18_eus/step/skip_bny.rb]: Removed

File: lib/engine/game/g_18_eus/game.rb (hunk 18 above)
Comment (outsidePasser) [reply lib/engine/game/g_18_eus/game.rb]: Also removed this method because never needed to get stock_movement and multiplier at the same time - so now have two methods, one for each.

File: lib/engine/game/g_18_eus/step/dividend.rb (hunk 24 above)
Comment (outsidePasser) [reply lib/engine/game/g_18_eus/step/dividend.rb]: Didn't need to add it - can check in game for current_entity being BNY.

File: lib/engine/game/g_18_eus/step/dividend.rb (hunk 25 above)
Comment (outsidePasser) [reply lib/engine/game/g_18_eus/step/dividend.rb]: Moved to change_share_price, and changed to up.

File: lib/engine/game/g_18_eus/step/dividend.rb (hunk 26 above)
Comment (outsidePasser) [reply lib/engine/game/g_18_eus/step/dividend.rb]: Moved to change_share_price, and changed to up.

File: lib/engine/game/g_18_eus/step/buy_company.rb [hunk 33]
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
+require_relative '../../../step/buy_company'
+require_relative 'skip_bny'
+
+module Engine
+  module Game
+    module G18EUS
+      module Step
+        class BuyCompany < Engine::Step::BuyCompany
```
Comment (crericha) [inline lib/engine/game/g_18_eus/step/buy_company.rb]: Can you delete this file? We aren't going to use it.

File: lib/engine/game/g_18_eus/step/dividend.rb [hunk 34]
```diff
@@ -11,7 +11,26 @@ class Dividend < Engine::Step::Dividend
           DIVIDEND_TYPES = %i[payout half withhold].freeze
           include Engine::Step::HalfPay
 
+          ACTIONS = ['dividend'].freeze
+          def actions(entity)
+            return super unless entity == @game.bny
+
+            ACTIONS
```
Comment (crericha) [inline lib/engine/game/g_18_eus/step/dividend.rb:16]: Is this needed for the auto action to work?

File: lib/engine/game/g_18_eus/step/dividend.rb [hunk 35]
```diff
@@ -24,6 +43,35 @@ def share_price_change(entity, revenue = 0)
           def movement_str(times, dir)
             "#{times / 2} #{dir}"
           end
+
+          def corporation_dividends(entity, per_share)
+            entity == @game.bny ? 0 : super
+          end
+
+          def process_dividend(action)
```
Comment (crericha) [inline lib/engine/game/g_18_eus/step/dividend.rb:51]: return super unless action.entity == @game.bny

Comment (crericha) [reply lib/engine/game/g_18_eus/step/dividend.rb:51]: Then don't put the bny logic in an if block. Cleaner this way.

File: lib/engine/game/g_18_eus/step/dividend.rb [hunk 36]
```diff
@@ -24,6 +43,35 @@ def share_price_change(entity, revenue = 0)
...
+                player.spend(interest * player.loans, @game.bank)
+                @log << "#{player.name} pays #{@game.format_currency(interest * player.loans)} in interest " \
+                        "(#{@game.format_currency(interest)} per share)"
+              end
+            end
+
+            super
+          end
+
+          def stock_movement_to_spaces(movement)
```
Comment (crericha) [inline lib/engine/game/g_18_eus/step/dividend.rb]: Does this make sense here or in the game class? No strong opinion other than my initial reaction was, huh, why is this here.

File: lib/engine/game/g_18_eus/step/loan_taker.rb [hunk 37]
```diff
@@ -0,0 +1,41 @@
+# frozen_string_literal: true
+
+module Engine
+  module Game
+    module G18EUS
+      module Step
+        module LoanTaker
+          def actions(entity)
+            return super unless entity == current_entity
```
Comment (crericha) [inline lib/engine/game/g_18_eus/step/loan_taker.rb]: actions = super.dup
return actions unless entity == current_entity

then don't call super again below.

**Review by crericha — APPROVED** [review]

File: lib/engine/game/g_18_eus/step/dividend.rb (hunk 34 above)
Comment (outsidePasser) [reply lib/engine/game/g_18_eus/step/dividend.rb:16]: Looks like it is not

File: lib/engine/game/g_18_eus/step/dividend.rb (hunk 35 above)
Comment (outsidePasser) [reply lib/engine/game/g_18_eus/step/dividend.rb:51]: Changed

File: lib/engine/game/g_18_eus/step/dividend.rb (hunk 36 above)
Comment (outsidePasser) [reply lib/engine/game/g_18_eus/step/dividend.rb]: It's only being used in this step - so I think maybe it belongs here.  Thinking this could just be a constant hash, tho.

File: lib/engine/game/g_18_eus/step/loan_taker.rb (hunk 37 above)
Comment (outsidePasser) [reply lib/engine/game/g_18_eus/step/loan_taker.rb]: Changed
